pkg/resourcemanager/controller/csrapprover/reconciler.go

```diff
@@ -242,8 +242,12 @@ func (r *Reconciler) mustApproveKubeAPIServerClient(ctx context.Context, csr *ce
 
 	switch {
 	case strings.HasPrefix(csr.Spec.Username, "system:bootstrap:"):
-		if _, found := machine.Labels[machinev1alpha1.NodeLabelKey]; found {
-			return fmt.Sprintf("Cannot use bootstrap token since gardener-node-agent for machine %q is already bootstrapped", machineName), csrDenied, nil
+		if nodeName, found := machine.Labels[machinev1alpha1.NodeLabelKey]; found {
+			if err := r.TargetClient.Get(ctx, client.ObjectKey{Name: nodeName}, &corev1.Node{}); err == nil {
+				return fmt.Sprintf("Cannot use bootstrap token since gardener-node-agent for machine %q is already bootstrapped", machineName), csrDenied, nil
+			} else if !apierrors.IsNotFound(err) {
+				return "", csrNoOpinion, fmt.Errorf("error getting node object with name %q: %w", nodeName, err)
+			}
```

Comment:
It turned out that depending on the MCM implementation the machine might have a node label before the bootstrapping is finished. In this case the CSR is denied even though GNA bootstrapping did not finish yet.

If we check for the existence of the node too, this cannot happen because kubelet creates the node and it starts after GNA finished bootstrapping.
I added this case to the integration test.